### PR TITLE
Support building with cygwin perl

### PR DESCRIPTION
**Pull request by yoe**
The match on the line just above the open, when using cygwin perl, does
not end with a / character. The result of that is that we try to find a
file "ms../crypto/perlasm/x86_64-xlate.pl", which obviously does not
exist.

Fix by adding an explicit / character

Signed-off-by: Wouter Verhelst w@uter.be

### CODE REVIEW

**Comment by levitte** [comment]
This is cygwin perl used from the Windows Command Prompt or similar, right?  What happens is that `${dir}` gets the value `ms\\`, and the command line string that's being opened becomes `| "/usr/bin/perl" ms\\../crypto/perlasm/x86_64-xlate.pl`, which is kinda sorta correct, except that Cygwin perl will massage `\\` in a way that doesn't work well in the end, resulting in the attempt to load ms../crypto/perlasm/x86_64-xlate.pl...

Your fix comes with a different problem, though, which you will see if you do this:

```
cd ms
perl uplink-x86_64.pl
```

then, it will suddenly try to load `/../crypto/perlasm/x86_64-xlate.pl`, and that's not what you intended.

Because this and other quirks we've found along the way, we *do not* support using cygwin perl in native Windows command line, just like we don't support the use of  ActiveState Perl in cygwin.

However, if you still want to fix this particular issue properly, I'd say adding this line rather than your slash has a greater chance of achieving something:

``` perl
$dir =~ s|\\|/|g;
```

Try it out.

**Comment by levitte** [comment]
Ping? @yoe, do you have any comment on my feedback?

**Comment by dot-asm** [comment]
Here is coherent argument for *not* supporting cygwin perl with nmake builds. One can actually argue that current way is kind of opportunistic and somewhat inadequate. Indeed, when it extracts $dir, it looks for backslashes and slashes, and then adds relative path with slashes. Yes, we are *so* lucky that Windows accepts both backslashes and slashes, and even in same pathname. But how would "non-opportunistic" code look like? It would have to invoke on File::Spec facility, right? Originally I didn't want to complicate things, but more importantly note that in such case cygwin perl wouldn't find $dir *at all*. Because script that invokes `uplink-*.pl` passes name with backslash while cygwin path delimiter is ... slash and it would remain "blind" to backslashes passed by caller. Also note that it means that it's possible to solve the problem from the other end, i.e. by modifying `ms/do_*.bat` files and replacing `perl ms\uplink*` with `perl ms/uplink*`. But then one would be able to say that such move would be ... opportunistic. Bottom line is that doing it right (with File::Spec) will make it impossible to use cygwin perl, hence it shouldn't be supported.

**Comment by dot-asm** [comment]
Side note. This is 1.0.2. But what happens in master context? There we don't have `ms/do_*.bat` scripts, and it is perl that generates commands in makefile. What happens if you let cygwin perl generate makefile for nmake? Well, it generates paths with *slashes*, except cases when backslash is explicitly used in windows-makefile.template. Just like any generic Windows program nmake accepts either and would actually make some progress. I haven't actually tested though, but it looks like it would even compile uplink things. But does for example `ENGINESDIR_dir=\OpenSSL/lib/engines-1_1` look right? `"$(PERL)" "$(SRCDIR)\test\run_tests.pl" $(TESTS)` [if $(PERL) is cygwin one]?

Bottom line, again, we are better off using "matching" perl. It's reasonable to require it and to refuse to support "non-matching" one. Though "matching" can be a bit misleading term. Truth be told It's more important that it matches *make* than "OS". This calls for following question. Wouldn't it be more natural to arrange it so that cygwin perl generates make-file for cygwin make, and cygwin make calls VCS compiler?

**Comment by yoe** [comment]
Sorry, forgot about this.

I noticed it didn't work, and that it did work with those updates, but I didn't know the rabbit hole was that deep. I've since switched to 1.1 and using mingw to build OpenSSL for Windows, so for me the issue is solved, and I can't spend more time debugging/fixing this.

I won't close the pull request since someone else might want to fix it, but if not, feel free to close it.

**Comment by levitte** [comment]
Closing seems to be the best bet.  If someone else wants to make a similar complaint, (s)he will probably just create a new issue.
